controllers/finalizers.go

```diff
@@ -35,7 +34,11 @@ func reconcileFinalizers(ctx context.Context, r *SpecialResourceReconciler) erro
 }
 
 func finalizeNodes(ctx context.Context, r *SpecialResourceReconciler, remove string) error {
-	for _, node := range cache.Node.List.Items {
+	nodeList, err := r.KubeClient.GetNodesByLabels(ctx, r.specialresource.Spec.NodeSelector)
+	if err != nil {
+		return fmt.Errorf("Failed to get node list based on NodeSelector labels during finalization")
```

Comment:
How about keeping the underlying error's context?
```suggestion
		return fmt.Errorf("Failed to get node list based on NodeSelector labels during finalization: %v", err)
```

Reply:
fixed